xbatcher/testing.py

```diff
@@ -0,0 +1,236 @@
+    non_specified_ds_dims: Dict[Hashable, int],
+    non_input_batch_dims: Dict[Hashable, int],
+) -> int:
+    """
+    Return the expected length of the sample dimension.
+
+    Parameters
+    ----------
+    generator : xbatcher.BatchGenerator
+        The batch generator object.
+    non_specified_ds_dics : dict
+        Dict containing all dimensions in the input dataset that are not
+        in the input_dims or batch_dims attributes of the batch generator.
+    non_input_batch_dims : dict
+        Dict containing all dimensions in specified in batch_dims that are
+        not also in input_dims
+
+    Returns
+    -------
+    s : int
+        Expected length of the sample dimension
+    """
+    if generator.concat_input_dims:
+        batch_concat_dims = [
+            generator.ds.sizes.get(k)
+            // np.nanmax([v, generator.batch_dims.get(k, np.nan)])
+            for k, v in generator.input_dims.items()
+        ]
+    else:
+        batch_concat_dims = []
+    return int(
+        np.product(list(non_specified_ds_dims.values()))
+        * np.product(list(non_input_batch_dims.values()))
+        * np.product(batch_concat_dims)
+    )
+
+
+def get_batch_dimensions(generator: BatchGenerator) -> Dict[Hashable, int]:
+    """
+    Return the expected batch dimensions based on the ``input_dims``,
+    ``batch_dims``, and ``concat_input_dims`` attributes of the batch
+    generator.
+
+    Parameters
+    ----------
+    generator : xbatcher.BatchGenerator
+        The batch generator object.
+
+    Returns
+    -------
+    d : dict
+        Dict containing the expected dimensions for batches returned by the
+        batch generator.
+    """
+    # dimensions that are in the input dataset but not input_dims or batch_dims
+    non_specified_ds_dims = _get_non_specified_dims(generator)
+    # dimensions that are in batch_dims but not input_dims
+    non_input_batch_dims = _get_non_input_batch_dims(generator)
+    expected_sample_length = _get_sample_length(
+        generator=generator,
+        non_specified_ds_dims=non_specified_ds_dims,
+        non_input_batch_dims=non_input_batch_dims,
+    )
+    # input_dims stay the same, possibly with a new suffix
+    expected_dims = {
+        f"{k}_input" if generator.concat_input_dims else k: v
+        for k, v in generator.input_dims.items()
+    }
+    # Add a sample dimension if there's anything to get stacked
+    if (
+        generator.concat_input_dims
+        and (len(generator.ds.sizes) - len(generator.input_dims)) == 0
+    ):
+        expected_dims = {**{"input_batch": expected_sample_length}, **expected_dims}
+    elif (
+        generator.concat_input_dims
+        or (len(generator.ds.sizes) - len(generator.input_dims)) > 1
+    ):
+        expected_dims = {**{"sample": expected_sample_length}, **expected_dims}
+    else:
+        expected_dims = {
+            **non_specified_ds_dims,
+            **non_input_batch_dims,
+            **expected_dims,
+        }
```

Comment:
```suggestion
        expected_dims = dict(
            **non_specified_ds_dims,
            **non_input_batch_dims,
            **expected_dims,
        )
```

This will catch the unlikely case where a dim is found in multiple dim categories.